com.unity.shadergraph/Editor/ShaderFoundry/Linker/Properties/PropertyDeclarations.cs

```diff
@@ -0,0 +1,382 @@
+            {"UnityTexture3D", HandleUnityTexture3D },
+            {"VTPropertyWithTextureType", HandleVirtualTexture },
+        };
+
+        static Dictionary<string, Action<ShaderAttribute, FieldPropertyContext>> AttributeHandlers = new Dictionary<string, Action<ShaderAttribute, FieldPropertyContext>>
+        {
+            {"Gamma", DefaultAttributeProcessor},
+            {"HDR", DefaultAttributeProcessor},
+            {"HideInInspector", DefaultAttributeProcessor},
+            {"MainTexture", DefaultAttributeProcessor},
+            {"MainColor", DefaultAttributeProcessor},
+            {"NoScaleOffset", DefaultAttributeProcessor},
+            {"Normal", DefaultAttributeProcessor},
+            {"PerRendererData", DefaultAttributeProcessor},
+            {"KeywordEnum", DefaultAttributeProcessor},
+            {"Toggle", DefaultAttributeProcessor},
+            {"ToggleOff", DefaultAttributeProcessor},
+            {"Enum", DefaultAttributeProcessor},
+            {"PowerSlider", DefaultAttributeProcessor},
+            {"IntRange", DefaultAttributeProcessor},
+            {"Space", DefaultAttributeProcessor},
+            {"Header", DefaultAttributeProcessor},
+            {"VTexLayers", DefaultAttributeProcessor },
+        };
+
+        internal static FieldPropertyData Extract(ShaderType type, string name, IEnumerable<ShaderAttribute> attributes)
+        {
+            var propertyAttribute = PropertyAttribute.FindFirst(attributes);
+            if (propertyAttribute == null)
+                return null;
+
+            if (!TypeHandlers.TryGetValue(type.Name, out var handler))
+                throw new Exception($"Type {type.Name} is an invalid property type.");
+
+            var container = type.Container;
+            var context = new FieldPropertyContext(container, type, name, attributes, propertyAttribute);
+            ProcessAttributes(context, attributes);
+
+            var result = new FieldPropertyData();
+            handler(context, result);
+            return result;
+        }
+
+        static void ProcessAttributes(FieldPropertyContext context, IEnumerable<ShaderAttribute> attributes)
+        {
+            if (attributes == null)
+                return;
+            foreach (var attribute in attributes)
+            {
+                if (AttributeHandlers.TryGetValue(attribute.Name, out var attributeHandler))
+                    attributeHandler(attribute, context);
+            }
+        }
+
+        static void DefaultAttributeProcessor(ShaderAttribute attribute, FieldPropertyContext context)
+        {
+            if (!context.ProcessedAttributes.ContainsKey(attribute.Name))
+                context.ProcessedAttributes.Add(attribute.Name, attribute);
+        }
+
+        static bool HandleFloatType(FieldPropertyContext context, FieldPropertyData result)
+        {
+            var displayType = GetDisplayType(context, "Float", new HashSet<string> { "Integer", "Int", "Range" });
```

Comment:
Why is HandleFloatType handling Integer as well?
It's backed by a "int32"

Reply:
This is to handle when you have a float property but want it to have the "Integer" property type. From what I saw this is a supported thing unless I misunderstood.

Reply:
This is handling fields marked as `[Property]` based upon the field type and extracting the relevant uniform and material properties.

Reply:
So it's a `float` uniform here, right?
Yes, that's supported then :)

Reply:
Yes, I guess I didn't say that explicitly, but the field type is used to determine the uniform type whereas the material property type can be configured